metriq_gym/quantinuum/job.py

```diff
@@ -19,36 +66,121 @@ def _get_ref(self):
+            self._ref_obj = None  # Clear cache to get new ref
+            return
+
+        # Check if compilation is complete
+        compile_ref = self._get_ref()
+        compile_status = getattr(compile_ref, "last_status", None)
+
+        if compile_status != "COMPLETED":
+            raise RuntimeError(
+                f"Cannot get results - compilation not yet complete (status: {compile_status}). "
+                f"Check job.status() or job.phase to monitor compilation progress."
+            )
+
+        # Get compiled circuits and create execute job
+        from qnexus.models.language import Language
+
+        compiled_refs = [item.get_output() for item in qnx.jobs.results(compile_ref)]
+        execute_job = qnx.start_execute_job(
+            programs=compiled_refs,
+            name=self._execute_name_prefix,
+            n_shots=[self._shots] * len(compiled_refs),
+            backend_config=self._backend_config,
+            project=self._project,
+            language=Language.QIR,
+        )
+
+        self._execute_job_id = (
+            getattr(execute_job, "id", None)
+            or getattr(execute_job, "job_id", None)
+            or str(execute_job)
+        )
+
+        # Switch to tracking the execute job
+        self.id = self._execute_job_id
+        self._is_compile_job = False
+        self._ref_obj = None  # Clear cache
+
     def result(self) -> Result:
+        # Return cached result if available
+        if self._cached_result is not None:
+            return self._cached_result
+
+        # Ensure we're tracking the execute job, not compile job
+        self._ensure_execute_job()
+
         ref = self._get_ref()
         results = qnx.jobs.results(ref)
-        if not results:
-            # TODO: don't make it blocking
-            qnx.jobs.wait_for(ref)
-            results = qnx.jobs.results(ref)
         if not results:
             raise RuntimeError(f"No results available for job {self.id}")
 
-        counts = results[0].download_result().get_counts()
-        norm_counts = {"".join(map(str, k)): v for k, v in counts.items()}
+        # Collect counts from all results (handles batch job submissions)
+        # Always return a list for consistency, even with single circuit
+        all_counts = []
+        for result in results:
+            counts = result.download_result().get_counts()
+            norm_counts = {"".join(map(str, k)): v for k, v in counts.items()}
+            all_counts.append(norm_counts)
 
-        return Result(
+        # Cache the result for future calls
+        self._cached_result = Result(
             device_id=self._device.name if self._device else "unknown",
             job_id=self.id,
             success=True,
-            data=GateModelResultData(measurement_counts=norm_counts),
+            data=GateModelResultData(measurement_counts=all_counts),
         )
+        return self._cached_result
 
     def status(self) -> JobStatus:
```

Comment:
This is the status of either the compile job or the execution job, depending on the phase?

If so, are there any assumptions on the status transitions? For example, previously with just an execution job, I assume once you get to say RUNNING, you don't transition back to QUEUED or INITIALIZING. But in this case, could the compile job go from RUNNING to COMPLETED and then back to QUEUED for the execution job?